### ISSUE TITLE
User tags broken

### ISSUE THREAD

**Issue by mjuric** [RobertLuptonTheGood/eups#93]
User tags declared in `~/.eups` are forgotten when `setups.sh` is run again. This was first noticed in the [discussion on discourse](https://community.lsst.org/t/change-of-behavior-in-eups/523). I've managed to reproduce the bug and isolate its introduction to commit babde22. More specifically, [it's this line](https://github.com/RobertLuptonTheGood/eups/commit/babde226f096d20fcea56b069cf63c31f04acd66#diff-d040f65ada06eb8708ac6a26429f9c00R126) that triggers it.

Based on a first look, I suspect it has something to do with a) user tags only being stored in the pickled product DB (i.e., in `~/.eups/ups_db/DarwinX86.pickleDB1_3_0`) -- which may be the real bug -- and b) the line mentioned above triggering a rebuild of that database, which didn't use to happen before.

Those pickle databases were always something I tried to avoid touching as the code seems to be quite messy, so this may be a task for someone who's more intimately familiar with them ( @RobertLuptonTheGood )? I suspect the real bug is in the fact that the user tag isn't getting written out to a `.chain` file in `~/.eups`, but I don't know enough about how user tags are implemented and meant to work to be certain.

**Comment by mjuric** [comment]
Use the script in [this gist](https://gist.github.com/mjuric/3e86b3896018376544d2#file-testusertag-sh) to reproduce the problem. Copy it into the root of your EUPS source, and run something like:
```
(git reset --hard && git checkout master && ./configure --prefix=$PWD --with-eups=$PWD/stack) && ./testUserTag.sh
```
The output will be:
```
[mjuric@gallifrey (master) ~/projects/eups]$ (git reset --hard && git checkout master && ./configure --prefix=$PWD --with-eups=$PWD/stack) && ./testUserTag.sh
HEAD is now at 097f142 Merge pull request #83 from RobertLuptonTheGood/enable-travis
Already on 'master'
Your branch is up-to-date with 'origin/master'.
configure: creating ./config.status
config.status: creating Makefile
config.status: creating bin/eups_impl.py
config.status: creating bin/eups_setup_impl.py
config.status: creating bin/eups
config.status: creating bin/eups_setup
config.status: creating ups/eups.table
Writing a csh startup script
Writing a sh startup script
Linking zsh/dash startup script from sh one for historical reasons
make: `git.version' is up to date.
Creating user data directory: /Users/mjuric/.eups
eups list: No products found
EUPS Version: 2.0.1-7-g097f142
Warning: path /Users/mjuric/projects/eups/tmpstack/DarwinX86/prod/a is absolute, not relative to EUPS_PATH
After declaring the product with user tag:
prod                  tag:mjuric 	mjuric
Parent shell, w/o calling setups.sh:
prod                  tag:mjuric 	mjuric
Sibling shell, after calling setups.sh:
prod                  tag:mjuric
```
(note how the tag is missing on the last line).

**Comment by r-owen** [comment]
Until this is fixed, here is a workaround: https://community.lsst.org/t/change-of-behavior-in-eups/523/3

**Comment by danielsf** [comment]
If I can add urgency to this issue: DESC users are starting to run into this bug.  It is making it very hard for us (the Sims team, mostly) to coach them through using EUPS and the stack.

**Comment by mjuric** [comment]
@danielsf, @r-owen -- could you try the fix I comitted to the `user-tags-issue-workaround` branch?

[It's simple enough](https://github.com/RobertLuptonTheGood/eups/commit/95d6e7f79fd3d84f43eab36e3bcb1a8789599c19) you could even apply it by hand to your installed EUPS and see if it made a difference. If this solves the problem, I could cut a new version ASAP

**Comment by danielsf** [comment]
I'm trying, but I cannot seem to get the setup script.  For example, after running
```
make install
```
I see
```
Remember to source /Users/danielsf/physics/test_eups/bin/setups.{c,z,}s
```
However,
```
/Users/danielsf/physics/test_eups/bin/
```
is empty

**Comment by mjuric** [comment]
Did you run `configure`?

**Comment by danielsf** [comment]
Yes.
```
./configure --prefix=/Users/danielsf/physics/test_eups
```
I did not pass use
```
--with-eups
```
because I wanted to declare packages by hand

**Comment by danielsf** [comment]
just ran configure and make again and it worked

testing the patch now

**Comment by danielsf** [comment]
Okay.  It looks like your fix works

**Comment by mjuric** [comment]
Thanks, @danielsf, I merged the fix to master.

@r-owen, when you get a chance, could you also test it (from master branch)? If it works for you as well, I'll put out a new release.

**Comment by r-owen** [comment]
I installed eups  a81391f and it fixes the bug. I tested as follows:

- I made ~/.eups/ups_db writable 
- with the old eups, in my `pipe_base` git repo home directory I undeclared `pipe_base git`, then did `eups declare -r . pipe_base git -t rowen` and, finally, started a new terminal and eups and did `eups list pipe_base`. The git version was declared but the `rowen` tag was missing.
- I repeated the above with eups master a81391f and it worked as expected: the `rowen` tag was still present

One thing: I have one failing unit test:

    localhost$ rm -r tests/ups_db/base/
    localhost$ make test
    Running tests
    python tests/testAll.py
    ....................................................................................................................
    ----------------------------------------------------------------------
    Ran 116 tests in 6.984s

    OK
    Running server tests (requires network)
    python tests/testServerAll.py
    The remote server tests rely on a server that used to run on lsstcorp.org; skipping
    .E..............
    ======================================================================
    ERROR: testGetTagNames (testServerLocal.LocalDistribServerTestCase)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Users/rowen/UW/LSST/eups_git/tests/testServerLocal.py", line 114, in testGetTagNames
        tags = DistribServer.getTagNames(self.ds)
      File "/Users/rowen/UW/LSST/lsstsw/eups/git/python/eups/distrib/server.py", line 122, in getTagNames
        tagNames.append(x[:-5])
    NameError: global name 'x' is not defined

    ----------------------------------------------------------------------
    Ran 16 tests in 0.164s

    FAILED (errors=1)
    make: *** [test] Error 1
    localhost$

**Comment by mjuric** [comment]
@r-owen , that should be fixed by 062a1b30958088735d7822346d323a37260f97cc (to be merged soon).

PS: I'm closing the issue given the tags seem to work again.

**Comment by boutigny** [comment]
Is there a standard procedure to update eups in the lsstsw directory tree ?
